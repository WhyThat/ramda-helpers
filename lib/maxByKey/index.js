import { curry, maxBy, prop, reduce } from 'ramda';

const reducer = propStr => maxBy(prop(propStr));

const maxByKey = curry((propStr, defaultObj, list) => reduce(reducer(propStr), defaultObj, list));

/**
 * Returns the object which have the greater value by the key, * otherwise returns default object
 * @name maxByKey
 * @param {*} defaultObject
 * @param {Array}
 * @returns {*}
 * @example
 * maxByKey('a', { a: 0 }, [{ a: 6 }, { a: 5 }]) //=> { a: 6 }
 */
export default maxByKey;
