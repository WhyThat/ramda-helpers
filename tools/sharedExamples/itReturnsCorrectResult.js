export default (result) => {
  test('returns correct result', () => {
    expect(result).toMatchSnapshot();
  });
};
