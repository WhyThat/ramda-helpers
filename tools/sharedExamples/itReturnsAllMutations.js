import itReturnsCorrectResult from './itReturnsCorrectResult';

export default (fn, mutations) => {
  mutations.forEach((mutation, index) => {
    const name = mutation.name || `mutations ${index}`;

    describe(name, () => {
      itReturnsCorrectResult(fn(mutation.props));
    });
  });
};
