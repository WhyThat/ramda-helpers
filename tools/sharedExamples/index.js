
import itReturnsAllMutations from './itReturnsAllMutations';
import itReturnsCorrectResult from './itReturnsCorrectResult';

export {
  itReturnsAllMutations,
  itReturnsCorrectResult,
};
