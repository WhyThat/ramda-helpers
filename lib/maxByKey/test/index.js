import { itReturnsCorrectResult } from '../../test/sharedExamples';
import maxByKey from '../';

describe('maxByKey', () => {
  const mutations = [{
    name: 'With null',
    props: [null, null, []],
  }, {
    name: 'key not defined in nested object',
    props: ['b', { a: 0 }, []],
  }, {
    name: 'key defined',
    props: ['a', { a: 0 }, [{ a: 6 }, { a: 5 }]],
  }];

  mutations.forEach((mutation) => {
    describe(mutation.name, () => {
      itReturnsCorrectResult(maxByKey(...mutation.props));
    });
  });
});
