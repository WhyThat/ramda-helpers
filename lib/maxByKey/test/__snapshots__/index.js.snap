// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`maxByKey With null returns correct result 1`] = `null`;

exports[`maxByKey key defined returns correct result 1`] = `
Object {
  "a": 6,
}
`;

exports[`maxByKey key not defined in nested object returns correct result 1`] = `
Object {
  "a": 0,
}
`;
